File: src/math/line-segment.ts
import Line from "./line";
import { compare } from "./number";
import Rectangle from "./rectangle";
import Vector2 from "./vector2";

class LineSegment {
  public readonly a: Vector2;
  public readonly b: Vector2;

  private _line?: Line;
  private _boundingRectangle?: Rectangle;

  public get line(): Line {
    if (this._line === undefined) {
      this._line = this.computeLine();
    }

    return this._line;
  }

  public get boundingRectangle(): Rectangle {
    if (this._boundingRectangle === undefined) {
      this._boundingRectangle = this.computeBoundingRectangle();
    }

    return this._boundingRectangle;
  }

  public constructor(a: Vector2, b: Vector2) {
    this.a = a;
    this.b = b;
  }

  private computeLine(): Line {
    return Line.fromPoints(this.a, this.b);
  }

  private computeBoundingRectangle(): Rectangle {
    const [minX, maxX] = [this.a.x, this.b.x].sort(compare);
    const [minY, maxY] = [this.a.y, this.b.y].sort(compare);

    return new Rectangle(minX, minY, maxX - minX, maxY - minY);
  }
}

export default LineSegment;

File: logic/src/math/line_segment.rs
use once_cell::unsync::OnceCell;

use super::{order, Line, Rectangle, Vector2};

#[derive(Debug)]
pub struct LineSegment {
    a: Vector2<f32>,
    b: Vector2<f32>,
    line: OnceCell<Line>,
    bounding_rectangle: OnceCell<Rectangle<f32>>,
}

impl LineSegment {
    pub fn new(a: Vector2<f32>, b: Vector2<f32>) -> Self {
        Self {
            a,
            b,
            line: OnceCell::new(),
            bounding_rectangle: OnceCell::new(),
        }
    }

    pub fn a(&self) -> &Vector2<f32> {
        &self.a
    }

    pub fn b(&self) -> &Vector2<f32> {
        &self.b
    }

    pub fn line(&self) -> &Line {
        if self.line.get().is_none() {
            _ = self.line.set(Line::from_points(self.a, self.b));
        }

        self.line.get().unwrap()
    }

    pub fn bounding_rectangle(&self) -> &Rectangle<f32> {
        if self.bounding_rectangle.get().is_none() {
            let (min_x, max_x) = order((self.a.x, self.b.x));
            let (min_y, max_y) = order((self.a.y, self.b.y));

            _ = self.bounding_rectangle.set(Rectangle::new(
                min_x,
                min_y,
                max_x - min_x,
                max_y - min_y,
            ));
        }

        self.bounding_rectangle.get().unwrap()
    }
}
